SO/Ficha2.md

```diff
@@ -0,0 +1,290 @@
+- Ficheiros temporários são removidos.
+
+Finalmente `exit` chama `_exit`. Por isso, `_exit` é mais abrupta enquanto que
+`exit` termina o processo normalmente.
+
+### Wait, zombies e órfãos
+Criar processos filhos vem com responsabilidades da parte do processo pai.
+
+Para esperar que um processo filho termine temos as funções:
+```c
+pid_t wait(int* status);
+pid_t waitpid(pid_t pid, int* status, int options);
+```
+Ambas retornam o `pid` do processo que terminou, e colocam no `status` o *exit
+status* desse processo. Em caso de erro `-1` é retornado.
+
+Enquanto que o `wait` retorna mal encontre um processo filho que tenha
+terminado.
+
+O `waitpid` por outro lado, permite um ter mais controlo sobre quando retornar.
+
+O parâmetro `pid` pode tomar valores que não `pid`s, mas para além de `-1` (que
+significa "qualquer processo"), estes saem fora da matéria da cadeira
+[<sup>\[2\]</sup>](processGroups).
+
+Ao parâmetro `options` pode ser passado `0` para não ativar nenhuma opção ou
+as opções especificadas na *man page*, das quais a mais útil é `WNOHANG` que
+permite ao `waitpid` não bloquear caso nenhum processo filho tenha terminado.
+
+O que acontece ao processo filho se não fizermos wait?
+
+Enquanto o processo pai existir os processos filhos que já terminaram são
+designados de *zombies*, o sistema operativo mantém apenas a informação
+necessária para que o processo possa ser *waited for* mais tarde. Mesmo assim,
+enquanto o *zombie* existir uma entrada na tabela de processos continua a
+existir e se esta tabela encher não é possível criar novos processos.
+
+No caso de o pai terminar, normalmente, sem ter feito *wait* de todos os seus
+filhos, estes tornam se órfãos e passam a ser filhos do processo `init` que tem
+sempre `pid` 1.
+
+O `init` é o primeiro processo que arranca sempre que o sistema operativo
+inicia, tem sempre `pid` 1 e é o "pai de todos". Ele encarrega-se de
+periodicamente chamar o `wait` para todos os seus filhos de forma a evitar ter
+*zombies*, desta forma também serve como um *fall back plan* para processos
+órfãos não se tornarem *zombies* mais tarde.
+
+## Exit status
+
+Quando um programa termina retorna uma valor entre `0` e `255`, e o pai desse
+processo pode verificar esse número usando as macros `WIFEXITED` e
+`WEXITSTATUS`. Estas tem de ser usadas em conjunto para evitar *undefined
+behaviour*, sendo que `WEXITSTATUS` só pode ser usada caso `WIFEXITED` retornar
+`1` (`true`).
+
+```c
+int status;
+wait(&status);
+if(WIFEXITED(status)) {
+    puts("Child exited normally");
+    int exit_status = WEXITSTATUS(status);
+    printf("Exit status: %d\n", status);
+}
+```
+A razão para a ter de ser verificar se o processo terminou normalmente é que
+este pode ter terminado por receber um sinal, por exemplo, `SIGSEGV`
+(segmentation fault), e neste caso o valor que vem em `status` não é um exit
+status valido.
+
+## Exemplos
+
+### Procurar em paralelo.
+```c
+#include <stdio.h>
```

Comment:
let's unpack this.

includes devem estar separados por onde vêm, e depois por ordem alfabética.
por exemplo, de mais platform specific para menos:
```c
// Linux specific
#include <sys/types.h>
#include <sys/wait.h>

// POSIX specific
#include <unistd.h>

// C stdlib
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
```